File: leetcode/python/713_subarray_product_less_than_k.py
class Solution:
    def numSubarrayProductLessThanK(self, nums: list[int], k: int) -> int:
        l: int = 0
        res: int = 0
        prod: int = 1
        for i, n in enumerate(nums):
            prod *= n
            while l <= i and prod >= k:
                prod /= nums[l]
                l += 1
            res += i - l + 1
        return res

if __name__ == '__main__':
    s = Solution()
    print(s.numSubarrayProductLessThanK([10,5,2,6], 100))
    print(s.numSubarrayProductLessThanK([1,2,3], 0))

File: leetcode/rust/src/713_subarray_product_less_than_k.rs
impl Solution {
    pub fn num_subarray_product_less_than_k(nums: Vec<i32>, k: i32) -> i32 {
        let mut count = 0;
        let mut product = 1;
        let mut left = 0;
        let mut right = 0;

        for right in 0..nums.len() {
            product *= nums[right];
            while product >= k && left <= right {
                product /= nums[left];
                left += 1;
            }
            count += right - left + 1;
        }
        count as i32
    }
}